cmake/dependencies/common.cmake

```diff
@@ -38,53 +38,63 @@ else()
-    set(FFMPEG_PREPARED_BINARIES "${CMAKE_CURRENT_SOURCE_DIR}/third-party/build-deps/ffmpeg/windows-x86_64")
-elseif(APPLE)
-    if(CMAKE_SYSTEM_PROCESSOR STREQUAL "x86_64")
-        set(FFMPEG_PREPARED_BINARIES "${CMAKE_CURRENT_SOURCE_DIR}/third-party/build-deps/ffmpeg/macos-x86_64")
-    elseif(CMAKE_SYSTEM_PROCESSOR STREQUAL "arm64")
-        set(FFMPEG_PREPARED_BINARIES "${CMAKE_CURRENT_SOURCE_DIR}/third-party/build-deps/ffmpeg/macos-aarch64")
-    elseif(CMAKE_SYSTEM_PROCESSOR STREQUAL "powerpc")
-        message(FATAL_ERROR "PowerPC is not supported on macOS")
-    else()
-        message(FATAL_ERROR "Unsupported system processor:" ${CMAKE_SYSTEM_PROCESSOR})
+    if(EXISTS ${FFMPEG_PREPARED_BINARIES}/lib/libhdr10plus.a)
+        set(HDR10_PLUS_LIBRARY
+                ${FFMPEG_PREPARED_BINARIES}/lib/libhdr10plus.a)
     endif()
-elseif(UNIX)
-    set(FFMPEG_PLATFORM_LIBRARIES va va-drm va-x11 vdpau X11)
-    if(CMAKE_SYSTEM_PROCESSOR STREQUAL "x86_64")
-        list(APPEND FFMPEG_PLATFORM_LIBRARIES mfx)
-        set(FFMPEG_PREPARED_BINARIES "${CMAKE_CURRENT_SOURCE_DIR}/third-party/build-deps/ffmpeg/linux-x86_64")
-        set(CPACK_DEB_PLATFORM_PACKAGE_DEPENDS "libmfx1,")
-        set(CPACK_RPM_PLATFORM_PACKAGE_REQUIRES "intel-mediasdk >= 22.3.0,")
-    elseif(CMAKE_SYSTEM_PROCESSOR STREQUAL "aarch64")
-        set(FFMPEG_PREPARED_BINARIES "${CMAKE_CURRENT_SOURCE_DIR}/third-party/build-deps/ffmpeg/linux-aarch64")
-    elseif(CMAKE_SYSTEM_PROCESSOR STREQUAL "ppc64le" OR CMAKE_SYSTEM_PROCESSOR STREQUAL "ppc64")
-        set(FFMPEG_PREPARED_BINARIES "${CMAKE_CURRENT_SOURCE_DIR}/third-party/build-deps/ffmpeg/linux-powerpc64le")
-    else()
-        message(FATAL_ERROR "Unsupported system processor:" ${CMAKE_SYSTEM_PROCESSOR})
-    endif()
-endif()
-set(FFMPEG_INCLUDE_DIRS
-        ${FFMPEG_PREPARED_BINARIES}/include)
-if(EXISTS ${FFMPEG_PREPARED_BINARIES}/lib/libhdr10plus.a)
-    set(HDR10_PLUS_LIBRARY
-            ${FFMPEG_PREPARED_BINARIES}/lib/libhdr10plus.a)
-endif()
-set(FFMPEG_LIBRARIES
+    set(FFMPEG_LIBRARIES
+            ${FFMPEG_PREPARED_BINARIES}/lib/libavcodec.a
+            ${FFMPEG_PREPARED_BINARIES}/lib/libavutil.a
+            ${FFMPEG_PREPARED_BINARIES}/lib/libcbs.a
+            ${FFMPEG_PREPARED_BINARIES}/lib/libSvtAv1Enc.a
+            ${FFMPEG_PREPARED_BINARIES}/lib/libswscale.a
+            ${FFMPEG_PREPARED_BINARIES}/lib/libx264.a
+            ${FFMPEG_PREPARED_BINARIES}/lib/libx265.a
+            ${HDR10_PLUS_LIBRARY}
+            ${FFMPEG_PLATFORM_LIBRARIES})
+else()
+    set(FFMPEG_LIBRARIES
         ${FFMPEG_PREPARED_BINARIES}/lib/libavcodec.a
         ${FFMPEG_PREPARED_BINARIES}/lib/libavutil.a
         ${FFMPEG_PREPARED_BINARIES}/lib/libcbs.a
-        ${FFMPEG_PREPARED_BINARIES}/lib/libSvtAv1Enc.a
         ${FFMPEG_PREPARED_BINARIES}/lib/libswscale.a
-        ${FFMPEG_PREPARED_BINARIES}/lib/libx264.a
-        ${FFMPEG_PREPARED_BINARIES}/lib/libx265.a
-        ${HDR10_PLUS_LIBRARY}
```

Comment:
I'm a bit concerned about not including these libraries. What's the reason for removing them? Licensing?

Reply:
No, we're dynamically linking them via `FFMPEG_PLATFORM_LIBRARIES`, although this is also optional via Gentoo's USE flags. I have tested without them. If someone else wanted to statically link these, I suppose they could still provide them via `FFMPEG_PLATFORM_LIBRARIES`, using their full paths if necessary, but I haven't tried that.

Regarding HDR, I believe we (optionally) bake this support into the same libx265.so as the non-HDR support.

Reply:
Perhaps we should use a FindFFmpeg.cmake module instead?

- https://github.com/obsproject/obs-studio/blob/master/cmake/Modules/FindFFmpeg.cmake
- https://gitlab.kitware.com/vtk/vtk/-/blob/master/CMake/FindFFMPEG.cmake
- https://github.com/snikulov/cmake-modules/blob/master/FindFFmpeg.cmake
- https://sources.debian.org/src/acoustid-fingerprinter/0.6-6/cmake/modules/FindFFmpeg.cmake/

I think the one from OBS would be a preferred starting point (probably just works as is)

Reply:
I've only looked at the OBS one, but I don't think this would work. All these modules probably assume that the ffmpeg it finds is either a set of dynamic libraries or a set of entirely self-contained static libraries. They would not know that libx264 is also needed, for example. You'd also need to be careful not to pick up a vanilla system ffmpeg.

To be honest, I'm not a big CMake fan and I particularly don't like the find modules. pkg-config alone can handle this much better. In this case, it's a little awkward because you want this to work with your custom ffmpeg build out of the box, and a user could be building from any directory. To avoid needing to run something prior to CMake, you'd need to use `configure_file` to generate the right .pc file on the fly. Something like this:

```
Name: FFmpeg for Sunshine
Description:
Version: 0
Libs: -L@CMAKE_SOURCE_DIR@/third-party/ffmpeg/lib -lva -lva-drm -lx264 -x265
Cflags: -I@CMAKE_SOURCE_DIR@/third-party/ffmpeg/include
```

Gentoo could provide its own up front, and you could generate the above if an existing .pc file is not found. I'm not entirely sure this will work, as `configure_file` might not actually write the file early enough, but I'd be willing to give it a shot. Let me know what you think.

Reply:
I gave this a go, and it was looking quite good, but then fell apart when I tried cross-compiling. pkg-config can handle cross-compiling very well under Gentoo, but it gets upset if you try to use files outside of the cross environment like in this case.

If you still want to clean this up a bit, renaming your directories in build-deps could make it simpler:

```cmake
set(FFMPEG_PREFIX ${CMAKE_SOURCE_DIR}/third-party/build-deps/ffmpeg/${CMAKE_SYSTEM_NAME}-${CMAKE_SYSTEM_PROCESSOR})

if(NOT IS_DIRECTORY ${FFMPEG_PREFIX})
    message(FATAL_ERROR "Unsupported operating system (${CMAKE_SYSTEM_NAME}) and processor (${CMAKE_SYSTEM_PROCESSOR}) combination")
endif()
```

Reply:
> renaming your directories in build-deps could make it simpler

That's do-able.

Do you happen to know what these values would need to be?

`${CMAKE_SYSTEM_NAME}-${CMAKE_SYSTEM_PROCESSOR}`

Reply:
> when I tried cross-compiling

I'd love to have some guidance on this, or info added to the docs. It could possibly save us many hours of CI time for arm builds.

Reply:
I noticed that pkgconf (the modern pkg-config implementation) has a `--define-prefix` feature that might solve the issue I was having above. This might be good for Gentoo in general, so I'll look into that.

I believe the `${CMAKE_SYSTEM_NAME}-${CMAKE_SYSTEM_PROCESSOR}` pairs would be:

* Windows-AMD64
* Darwin-x86_64
* Darwin-arm64
* Linux-x86_64
* Linux-aarch64
* Linux-ppc64le (you were also matching ppc64, but it's not ABI-compatible with your binaries)

I see you've been doing ARM builds with QEMU. That hurts. I've done very little cross-compiling outside of Gentoo, and I don't know much about GitHub Actions, but maybe I can offer some advice. I'll get back to you on that.

Reply:
> I believe the ${CMAKE_SYSTEM_NAME}-${CMAKE_SYSTEM_PROCESSOR} pairs would be

Cool, I'll work on changing this over.

> I see you've been doing ARM builds with QEMU.

It's actually not too bad for the Debian docker images. Those take ~45 minutes, which is somewhat bearable. But for some reason the Fedora images take 4-5 hours. For the life of me I cannot determine why they are so much slower.

> I don't know much about GitHub Actions

It's really not much more than running shell/terminal commands. So if you know how to do it in Linux terminal, that's all I would need to try to incorporate it.

Reply:
You don't seem to have reorganised ffmpeg yet. Can we merge this in the meantime since it isn't dependent on that?

Reply:
Sorry for the delay, it's on my list of things to do.